# Used Car Market

[eBay Kleinanzeigen](https://www.ebay-kleinanzeigen.de/), a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website is a popular market place for used cars. 

## Aim

To get insights into the used car market in Germany and to find the various factors that affect the
resale value, by analyzing [eBay Kleinanzeigen](https://www.ebay-kleinanzeigen.de/).

## Dataset

The entire dataset can be found [here](https://data.world/data-society/used-cars-data). For this analysis we have sampled 50,000 data points that was prepared by [Dataquest](https://www.dataquest.io/) including simulating a less-cleaned version of the data. Below are the description of the columns present in the dataset.

* `dateCrawled`: When this ad was first crawled. All field-values are taken from this date.
* `name`: Name of the car.
* `seller`: Whether the seller is private or a dealer.
* `offerType`: The type of listing
* `price`: The price on the ad to sell the car.
* `abtest`: Whether the listing is included in an A/B test.
* `vehicleType`: The vehicle Type.
* `yearOfRegistration`: The year in which the car was first registered.
* `gearbox`: The transmission type.
* `powerPS`: The power of the car in PS.
* `model`: The car model name.
* `odometer`: How many kilometers the car has driven.
* `monthOfRegistration`: The month in which the car was first registered.
* `fuelType`: What type of fuel the car uses.
* `brand`: The brand of the car.
* `notRepairedDamage`: If the car has a damage which is not yet repaired.
* `dateCreated`: The date on which the eBay listing was created.
* `nrOfPictures`: The number of pictures in the ad.
* `postalCode`: The postal code for the location of the vehicle.
* `lastSeen`: When the crawler saw this ad last online.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1]:
# Importing NumPy and Pandas library.

import numpy as np
import pandas as pd

In [2]:
# Reading autos.csv file into Pandas.

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
# To find additional information about autos dataset.

autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
# To display the first 5 rows in autos dataset.

autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The following observations can be made about the dataset.

* The dataset contains 20 columns, most of which are strings.
* Some columns have null values, but none have more than ~20% null values.
* The column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we can't just replace spaces with underscores.

## Data Cleaning

Let's begin our cleaning process by converting column names to snakecase.

**Converting column names to snakecase**

Make the following change to column names.

* `yearOfRegistration` to `registration_year`
* `monthOfRegistration` to `registration_month`
* `notRepairedDamage` to `unrepaired_damage`
* `dateCreated` to `ad_created`
* `nrOfPictures` to `num_photos`
* The rest of the column names from camelcase to snakecase.

In [5]:
# To display all column names in autos dataset.

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# Renaming column names to snakecase.

autos.rename({
    "dateCrawled": "date_crawled", "offerType": "offer_type", "abtest": "ab_test", "vehicleType": "vehicle_type",
    "yearOfRegistration": "registration_year", "powerPS": "power_ps", "monthOfRegistration": "registration_month",
    "fuelType": "fuel_type", "notRepairedDamage": "unrepaired_damage", "dateCreated": "ad_created",
                "nrOfPictures": "num_photos", "postalCode": "postal_code", "lastSeen": "last_seen"
}, axis="columns", inplace=True)

In [7]:
# To check the renamed columns.

autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

In [8]:
# To display the first 5 rows.

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

> Thus the column names are converted to snakecase.

Let's look at the descriptive statistics for all columns.

In [9]:
# To look at descriptive statistics for all columns. This is useful in data cleaning process.

autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Our observations are as follows.
  
* The `num_photos` column looks odd, we'll need to investigate this further.
* Need to convert `price` and `odometer` columns to int type.
* Need to change language of `unrepaired_damage` and `seller` column from german to english.
* `registration_year` and `registration_month` column looks odd, needs further analysis.

**num_photos column**

In [10]:
# To find the various value counts in num_photos column.

autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

As all values in num_photos column is zero, we can drop this column.

In [11]:
# To drop num_photos column

autos = autos.drop("num_photos", axis="columns")
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 19 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

> `num_photos` column was droped successfully.

**price and odometer column**

In [12]:
# To find the format in price column.

autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

We will remove `$`, `,` from price column and convert it to int type.

In [13]:
# To remove unwanted characters and to convert to int type.

price = (autos["price"]
         .str.replace("$", "", regex=False)
         .str.replace(",", "", regex=False)
         .astype(int)
         )
price.unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

In [14]:
# Assigning temporary variable to dataset.

autos["price"] = price

`price` column is converted to integer type.

In [15]:
# To find the format in odometer column.

autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

We will remove `km`, `,` from odometer column and convert it to int type.

In [16]:
# To remove unwanted characters and to convert to int type.

odometer = (autos["odometer"]
            .str.replace("km", "", regex=False)
            .str.replace(",", "", regex=False)
            .astype(int)
            )
odometer.unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

In [17]:
# Assigning temporary variable to dataset.

autos["odometer"] = odometer

`odometer` column is converted to integer type. Let's rename `odometer` column to `odometer_km` for more clarity.

In [18]:
# To rename odometer column to odometer_km.

autos.rename({"odometer": "odometer_km"}, axis="columns", inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

> Thus `price` and `odometer` columns are converted to integer type. `odometer` column is renamed to `odometer_km`.

Let's analyse both columns using describe method for incorrect or misleading information.

In [19]:
# To find misleading or incorrect data points.

print(autos["price"].describe())
print("\n")
print(autos["odometer_km"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


The minimum value in `price` column is 0 and the maximum value is 100 million. As both these values are not realistic, we need to further analyze `price` column. The values in this columns are rounded.

The values in `odometer_km` column are rounded and most of the readings are above 125,000 km.

In [20]:
# To view the count of various values in price column - ascending order.

print("Least Price\n-----------")
print(autos["price"].value_counts().sort_index(ascending=True).head(20))
print("\n")
print("Max Price\n---------")
print(autos["price"].value_counts().sort_index(ascending=True).tail(20))

Least Price
-----------
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


Max Price
---------
197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price, dtype: int64


As seen from the above result, there are 1421 used cars with price marked as 0 euros and various used cars with price above 10,000,000 euros. 

Given that eBay is an auction site, there could legitimately be items where the opening bid is 1 euro. We will keep the 1 euro items, but remove anything above 350,000 euros, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [21]:
# To select rows between 1 and 350,000 in price column.

autos = autos[autos["price"].between(1, 350000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

> Thus incorrect values are removed from `price` column. 

**registration_month column**

In [22]:
# registration_month column distribution.

autos["registration_month"].describe()

count    48565.000000
mean         5.782251
std          3.685595
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

The minimum value in registration_month column is zero and its a wrong entry. The month value should be between 1 and 12. Let's investigate this column further.

In [23]:
# To find the occurance of values in registration_month column.

autos["registration_month"].value_counts().sort_index()

0     4480
1     3219
2     2937
3     5003
4     4036
5     4031
6     4271
7     3857
8     3126
9     3330
10    3588
11    3313
12    3374
Name: registration_month, dtype: int64

There are 4480 rows with registration month as zero. Let's remove these rows from the dataset.

In [24]:
# To remove rows with registration month as zero.

autos = autos[autos["registration_month"].between(1, 12)]
autos["registration_month"].value_counts().sort_index()

1     3219
2     2937
3     5003
4     4036
5     4031
6     4271
7     3857
8     3126
9     3330
10    3588
11    3313
12    3374
Name: registration_month, dtype: int64

> Thus incorrect values are removed from `registration_month` column.

**registration_year column** 

In [25]:
# registration_year column distribution.

autos["registration_year"].describe()

count    44085.000000
mean      2004.111739
std         46.511679
min       1800.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9000.000000
Name: registration_year, dtype: float64

The minumum and maximum value in registration_year column are 1800 and 9000 respectively. 1800 represent years before the invention of car and 9000 is in the future. This column needs more analysis.

In [26]:
# To find the occurance of values in registration_year column.

autos["registration_year"].value_counts().sort_index()

1800    2
1927    1
1929    1
1931    1
1934    1
       ..
2800    1
4800    1
5000    2
6200    1
9000    1
Name: registration_year, Length: 86, dtype: int64

The data crawling was done in 2016, so let's remove any registration year after 2016. Other than the 2 occurance in 1800, the rest of the data starts from the year 1927. So let's remove 1800 from the dataset.


In [27]:
# To remove incorrect rows from autos dataset.

autos = autos[autos["registration_year"].between(1927, 2016)]
autos["registration_year"].describe()

count    42655.000000
mean      2003.195921
std          6.939793
min       1927.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

> Thus incorrect values are removed from `registration_year` column.

**unrepaired_damage column**

To change the language to english. Let's see the values already existing in the column.

In [28]:
# To view the value counts in unrepaired_damage column.

autos["unrepaired_damage"].value_counts()

nein    32719
ja       3981
Name: unrepaired_damage, dtype: int64

In [29]:
# Mapping from german to english language.

correction = {"nein": "No", "ja": "Yes"}
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(correction)

Let's check whether the mapping was done using value_counts method as shown below.

In [30]:
# To verify the result.

autos["unrepaired_damage"].value_counts()

No     32719
Yes     3981
Name: unrepaired_damage, dtype: int64

> Thus `unrepaired_damage` column was converted to english language.

**seller column**

To change the language to english. Let's see the values already existing in the column.

In [31]:
# To view the value counts in unrepaired_damage column.

autos["seller"].value_counts()

privat        42654
gewerblich        1
Name: seller, dtype: int64

In [32]:
# Mapping from german to english language.

correction = {"privat": "private", "gewerblich": "commercial"}
autos["seller"] = autos["seller"].map(correction)

Let's check whether the mapping was done using value_counts method as shown below.

In [33]:
# To verify the result.

autos["seller"].value_counts()

private       42654
commercial        1
Name: seller, dtype: int64

> Thus `seller` column was converted to english language.

## Data Analysis

**Date columns analysis**

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary.

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [34]:
# To view the format in which data points appear.

autos[["date_crawled", "last_seen", "ad_created",
       "registration_month", "registration_year"]].head()

date_crawled            last_seen           ad_created  \
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00   
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00   
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00   
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00   
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

Let's first analyse `date_crawled` column. It represents the date when the ad was first crawled.

In [35]:
# To find the distribution of data points in date_crawled column. ordered in ascending order by date.

(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
 )

2016-03-05    0.025062
2016-03-06    0.014301
2016-03-07    0.036362
2016-03-08    0.033220
2016-03-09    0.033126
2016-03-10    0.032212
2016-03-11    0.032986
2016-03-12    0.036971
2016-03-13    0.015637
2016-03-14    0.036619
2016-03-15    0.033900
2016-03-16    0.029399
2016-03-17    0.031368
2016-03-18    0.012894
2016-03-19    0.034416
2016-03-20    0.038167
2016-03-21    0.037299
2016-03-22    0.032728
2016-03-23    0.032306
2016-03-24    0.029211
2016-03-25    0.031087
2016-03-26    0.032259
2016-03-27    0.030946
2016-03-28    0.034697
2016-03-29    0.033876
2016-03-30    0.033431
2016-03-31    0.031790
2016-04-01    0.034345
2016-04-02    0.035916
2016-04-03    0.039222
2016-04-04    0.036596
2016-04-05    0.013082
2016-04-06    0.003118
2016-04-07    0.001454
Name: date_crawled, dtype: float64

In [36]:
# To find the distribution of data points in date_crawled column. ordered in ascending order by distribution.

(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_values()
 )

2016-04-07    0.001454
2016-04-06    0.003118
2016-03-18    0.012894
2016-04-05    0.013082
2016-03-06    0.014301
2016-03-13    0.015637
2016-03-05    0.025062
2016-03-24    0.029211
2016-03-16    0.029399
2016-03-27    0.030946
2016-03-25    0.031087
2016-03-17    0.031368
2016-03-31    0.031790
2016-03-10    0.032212
2016-03-26    0.032259
2016-03-23    0.032306
2016-03-22    0.032728
2016-03-11    0.032986
2016-03-09    0.033126
2016-03-08    0.033220
2016-03-30    0.033431
2016-03-29    0.033876
2016-03-15    0.033900
2016-04-01    0.034345
2016-03-19    0.034416
2016-03-28    0.034697
2016-04-02    0.035916
2016-03-07    0.036362
2016-04-04    0.036596
2016-03-14    0.036619
2016-03-12    0.036971
2016-03-21    0.037299
2016-03-20    0.038167
2016-04-03    0.039222
Name: date_crawled, dtype: float64

The website was crawled every day from 2016-03-05 to 2016-04-07. The date_crawled column is uniformly distributed with each date not exceeding 4 percentage of the total data points. 

Now let's analyse `last_seen` column. It represents the date when the crawler saw the ad last online.

In [37]:
# To find the distribution of data points in last_seen column. ordered in ascending order by date.

(autos["last_seen"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
 )

2016-03-05    0.001102
2016-03-06    0.003939
2016-03-07    0.005392
2016-03-08    0.007268
2016-03-09    0.009659
2016-03-10    0.010597
2016-03-11    0.012331
2016-03-12    0.023655
2016-03-13    0.008557
2016-03-14    0.012519
2016-03-15    0.015872
2016-03-16    0.016294
2016-03-17    0.027617
2016-03-18    0.007385
2016-03-19    0.015450
2016-03-20    0.020302
2016-03-21    0.020537
2016-03-22    0.020537
2016-03-23    0.018239
2016-03-24    0.019505
2016-03-25    0.018614
2016-03-26    0.016645
2016-03-27    0.015168
2016-03-28    0.020443
2016-03-29    0.021944
2016-03-30    0.024288
2016-03-31    0.023514
2016-04-01    0.023374
2016-04-02    0.025062
2016-04-03    0.025038
2016-04-04    0.023678
2016-04-05    0.127300
2016-04-06    0.224358
2016-04-07    0.133818
Name: last_seen, dtype: float64

In [38]:
# To find the distribution of data points in last_seen column. ordered in ascending order by distribution.

(autos["last_seen"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_values()
 )

2016-03-05    0.001102
2016-03-06    0.003939
2016-03-07    0.005392
2016-03-08    0.007268
2016-03-18    0.007385
2016-03-13    0.008557
2016-03-09    0.009659
2016-03-10    0.010597
2016-03-11    0.012331
2016-03-14    0.012519
2016-03-27    0.015168
2016-03-19    0.015450
2016-03-15    0.015872
2016-03-16    0.016294
2016-03-26    0.016645
2016-03-23    0.018239
2016-03-25    0.018614
2016-03-24    0.019505
2016-03-20    0.020302
2016-03-28    0.020443
2016-03-22    0.020537
2016-03-21    0.020537
2016-03-29    0.021944
2016-04-01    0.023374
2016-03-31    0.023514
2016-03-12    0.023655
2016-04-04    0.023678
2016-03-30    0.024288
2016-04-03    0.025038
2016-04-02    0.025062
2016-03-17    0.027617
2016-04-05    0.127300
2016-04-07    0.133818
2016-04-06    0.224358
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

Nows let's have a look at `ad_created` column. It represents the date at which an ad was created.

In [39]:
# ad_created column distribution.

print("The number of dates in ad_created column: ",
      autos["ad_created"].unique().shape[0])
print("\n")
print(autos["ad_created"]
      .str[:10]
      .value_counts(normalize=True, dropna=False)
      .sort_index()
      )

The number of dates in ad_created column:  74


2015-06-11    0.000023
2015-08-10    0.000023
2015-09-09    0.000023
2015-11-10    0.000023
2015-12-05    0.000023
                ...   
2016-04-03    0.039386
2016-04-04    0.036924
2016-04-05    0.011886
2016-04-06    0.003212
2016-04-07    0.001289
Name: ad_created, Length: 74, dtype: float64


More than 90% of ads where created during the same time the data was crawled. Oldest listing was created 9 month prior to the data crawling period. 

Now let's analyse registration_month column.

In [40]:
# registration_month column distribution.

autos["registration_month"].value_counts(normalize=True).sort_index()

1     0.072793
2     0.066815
3     0.113422
4     0.091291
5     0.090892
6     0.096565
7     0.087258
8     0.070824
9     0.076029
10    0.081936
11    0.075700
12    0.076474
Name: registration_month, dtype: float64

Now let's analyse registration_year column.

In [41]:
# registration_year column distribution.

autos["registration_year"].value_counts(normalize=True).sort_values()

1952    0.000023
1943    0.000023
1957    0.000023
1948    0.000023
1954    0.000023
          ...   
2006    0.060392
2004    0.060649
2000    0.060814
1999    0.061282
2005    0.063252
Name: registration_year, Length: 77, dtype: float64

> 97.46% of listing on ebay Kleinanzeigen has a registration year of 1990 or later.

**Average price for each Brand**

Let's use the `brand` and `price` column to find the average price of a used car corresponding to a brand.

In [42]:
# To find the popularity of various brands in the dataset.

brand = autos["brand"].value_counts().sort_values(ascending=False)
brand_index = brand.index
print("The total number of brands: ", brand.shape[0])
print("")
print("Brand name versus the number of listed cars")
print("-------------------------------------------")
print(brand)

The total number of brands:  40

Brand name versus the number of listed cars
-------------------------------------------
volkswagen        8882
bmw               4778
opel              4432
mercedes_benz     4254
audi              3755
ford              2932
renault           1982
peugeot           1280
fiat              1085
seat               783
skoda              727
nissan             660
mazda              637
smart              616
citroen            589
toyota             567
hyundai            437
mini               399
sonstige_autos     398
volvo              396
mitsubishi         342
honda              334
kia                319
alfa_romeo         282
porsche            273
suzuki             256
chevrolet          245
chrysler           148
dacia              118
jeep               101
daihatsu           100
land_rover          93
subaru              86
saab                73
jaguar              67
daewoo              64
rover               56
trabant             43
lanci

Volkswagen, BMW, Opel, Mercedes Benz, Audi and Ford are the top six brands by the number of cars listed in the platform. Volkswagen is the most popular car brand and have 8,882 cars versus 4,778 cars listed by BMW in the second place. These top car manufacturers focus on making affordable cars for the general public and it will be evident when we compare the brand versus the average price of each car. 

> Volkswagen, BMW and Opel are the top three used car brands being sold on ebay Kleinanzeigen.

In [43]:
# To find the average price for each brand.

brand_and_avg_price = {}
for value in brand_index:
    brand_to_price = autos.loc[autos["brand"] == value, "price"]
    avg_price = brand_to_price.mean()
    brand_and_avg_price[value] = avg_price
avg_price = pd.DataFrame.from_dict(
    brand_and_avg_price, orient="index", columns=["avg_price"])
avg_price.sort_values(by="avg_price", ascending=False)

avg_price
porsche         47482.344322
land_rover      19685.956989
sonstige_autos  13775.801508
jaguar          12459.283582
jeep            12049.049505
mini            10744.754386
audi             9776.165113
mercedes_benz    8882.735778
bmw              8652.106111
chevrolet        6835.963265
skoda            6569.939477
kia              6117.087774
dacia            5967.457627
volkswagen       5752.935150
hyundai          5588.768879
toyota           5283.061728
volvo            5109.189394
nissan           4955.293939
seat             4568.993614
subaru           4406.034884
honda            4379.116766
mazda            4349.810047
suzuki           4277.386719
alfa_romeo       4250.255319
ford             3957.076739
citroen          3902.359932
smart            3700.426948
chrysler         3601.912162
mitsubishi       3594.991228
saab             3329.410959
peugeot          3221.411719
lancia           3184.860465
opel             3167.165162
fiat             2950.726267
lada             2712.782609
renault          2616.376387
trabant          2289.976744
daihatsu         1761.310000
rover            1707.000000
daewoo           1088.750000

As seen from the above result, the top five car manufacturers by listed numbers are not present in the top five of this new list. There is a significant drop in average price after the top few rows. The top manufacturers in this list exclusively makes cars for the wealthy. The average price of Porsche is twice than that of the second place holder land_rover. Volkswagen has dropped significantly in the list due to its focus on mass manufacturing of affordable cars. Audi, Mercedes Benz and BMW are the only brands to maintain top ten spots in both lists. This makes it clear that these brands focus on both wealthy and the general economy. We observed that in the top 6 car brands, there's a distinct price gap.

* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between

> porsche, land_rover and sonstige_autos are the top three most priced used car brands on ebay Kleinanzeigen.

**Average price and kilometers traveled for top 6 manufacturers**

For the top 6 brands, let's use aggregation to understand the average kilometers traveled for those cars and if there's any visible link with mean price. 

In [44]:
# selecting the top six brands.

top_six_brands = brand.head(6)
top_six = top_six_brands.to_frame("count")
top_six["mean_price"] = avg_price
top_six

count   mean_price
volkswagen      8882  5752.935150
bmw             4778  8652.106111
opel            4432  3167.165162
mercedes_benz   4254  8882.735778
audi            3755  9776.165113
ford            2932  3957.076739

In [45]:
# To find the mean kilometers traveled by each brand.

brand_and_avg_km = {}
for value in top_six_brands.index:
    avg_km_each_brand = autos.loc[autos["brand"]
                                  == value, "odometer_km"].mean()
    brand_and_avg_km[value] = avg_km_each_brand
avg_km = pd.Series(brand_and_avg_km)
print("The average kilometers traveled by each car brand. \n")
print(avg_km)

The average kilometers traveled by each car brand. 

volkswagen       128183.967575
bmw              132489.535370
opel             128752.256318
mercedes_benz    130927.362482
audi             128593.874834
ford             124009.208731
dtype: float64


Let's try to combine the above series `avg_km` with `top_six` dataframe for more detailed view.

In [46]:
# Adding a new column mean_km to top_six dataframe.

top_six["mean_km"] = avg_km
top_six.sort_values(by="mean_price", ascending=False)

count   mean_price        mean_km
audi            3755  9776.165113  128593.874834
mercedes_benz   4254  8882.735778  130927.362482
bmw             4778  8652.106111  132489.535370
volkswagen      8882  5752.935150  128183.967575
ford            2932  3957.076739  124009.208731
opel            4432  3167.165162  128752.256318

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% of the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.


**Most common brand/model combinations**

Let's find out the most popular brand model combination on ebay Kleinanzeigen.

In [47]:
# To form a series that contains both brand and model.

brand_model = autos["brand"].str.strip() + " " + autos["model"].str.strip()

print("The total number of brand model combination is ",
      brand_model.value_counts().shape[0])
print("")
print("brand model versus total count")
print("------------------------------")
print(brand_model.value_counts().sort_values(ascending=False).head(20))

The total number of brand model combination is  290

brand model versus total count
------------------------------
volkswagen golf           3368
bmw 3er                   2427
volkswagen polo           1410
opel corsa                1403
volkswagen passat         1245
opel astra                1218
audi a4                   1137
mercedes_benz c_klasse    1082
bmw 5er                   1071
mercedes_benz e_klasse     914
audi a3                    782
audi a6                    750
ford focus                 704
ford fiesta                658
volkswagen transporter     612
peugeot 2_reihe            572
renault twingo             543
smart fortwo               515
mercedes_benz a_klasse     510
bmw 1er                    509
dtype: int64


The above list contains 20 most popular brand model combination. 

> Thus `Volkswagen Golf` is the most popular used car on ebay Kleinanzeigen.

**Price based on unrepaired damages**

Let's find out how cheaper are cars with damage than their non-damaged counterparts. The columns to consider for this analysis are `price` and `unrepaired_damage`.

In [48]:
# To find the value counts in unrepaired_damage column.

autos["unrepaired_damage"].value_counts(dropna=False)

No     32719
NaN     5955
Yes     3981
Name: unrepaired_damage, dtype: int64

NaN values are ignored for this analysis. Average price for damage and non-damaged used cars are calculated below.

In [49]:
# Average price for damage and non-damaged used cars.

avg_price_damage = autos.loc[autos["unrepaired_damage"]
                             == "Yes", "price"].mean()
avg_price_no_damage = autos.loc[autos["unrepaired_damage"]
                                == "No", "price"].mean()

print("The average price for non damaged used car is ",
      avg_price_no_damage, " euros.")
print("")
print("The average price for damaged used car is ", avg_price_damage, " euros.")

The average price for non damaged used car is  7296.379015251077  euros.

The average price for damaged used car is  2404.8836975634263  euros.


There is a huge difference in average price for damaged versus non-damaged used cars. The average price for a damaged used car is 2404.8837 versus 7296.379 euros for non damaged car.

> A damaged used car is ~ 67% cheaper than a non-damaged car.

**Kilometers traveled versus average price**

Let's check whether the total number of kilometers traveled by a car affect its average price. Columns used in this analysis are `odometer_km` and `price`.

In [50]:
# To find the value counts for odometer_km column.

odometer_km = autos["odometer_km"].value_counts(dropna=False)
print(odometer_km)

150000    27071
125000     4585
100000     1922
90000      1594
80000      1324
70000      1138
60000      1089
50000       956
40000       777
30000       736
20000       715
5000        527
10000       221
Name: odometer_km, dtype: int64


Lets find the average price for each of the above traveled kilometers by used car.

In [51]:
# To find the average price versus the odometer reading in used car.

avg_price_km = {}
for values in odometer_km.index:
    avg_price_each_km = autos.loc[autos["odometer_km"]
                                  == values, "price"].mean()
    avg_price_km[values] = avg_price_each_km
avg_price_odometer = pd.Series(avg_price_km)
print(avg_price_odometer.sort_values(ascending=False))

10000     21983.882353
20000     19053.556643
30000     16942.824728
40000     15559.854569
50000     14140.441423
60000     12666.706152
5000      11745.740038
70000     11188.612478
80000      9953.735650
90000      8717.777290
100000     8434.032258
125000     6389.061941
150000     3966.888811
dtype: float64


When the number of traveled kilometer incresed, the average price decreased as shown in the above result. The only exception to this rule is the 5000 km travled used car. It is priced equivalent to a car that has ~65,000 km reading. 5000 km is the lowest possible odometer reading in the dataset. There is a possibility that the car might have issues, as the seller is selling the car after such a low odometer reading. This explains the average pricing for such cars. 

> As the number of traveled kilometer increased, the average price decreased. Thus kilometers traveled by a used car is indirectly proportional to the average price.

## Conclusion

The following conclusions can be made about the used car market in Germany.

* Volkswagen, BMW and Opel are the most popular used cars for sale. These brands offer affordable cars that can be easily resold on eBay Kleinanzeigen.
* Porsche, Land Rover and Sonstige autos are the top three most priced used car brands. These brands tend to bring in a premium and thus can be listed at a higher sticker price.
* Damaged and kilometers traveled are the main reasons for a reduced price in used car.  